robosat/tools/train.py

```diff
@@ -73,39 +85,75 @@ def main(args):
     if model["common"]["cuda"]:
         torch.backends.cudnn.benchmark = True
 
+    weight = torch.Tensor(dataset["weights"]["values"])
+    optimizer = Adam(net.parameters(), lr=model["opt"]["lr"], weight_decay=model["opt"]["decay"])
+
+    resume = 0
     if args.checkpoint:
 
         def map_location(storage, _):
             return storage.cuda() if model["common"]["cuda"] else storage.cpu()
 
         # https://github.com/pytorch/pytorch/issues/7178
         chkpt = torch.load(args.checkpoint, map_location=map_location)
-        net.load_state_dict(chkpt)
-
-    optimizer = Adam(net.parameters(), lr=model["opt"]["lr"], weight_decay=model["opt"]["decay"])
-
-    weight = torch.Tensor(dataset["weights"]["values"])
-
-    criterion = CrossEntropyLoss2d(weight=weight).to(device)
-    # criterion = FocalLoss2d(weight=weight).to(device)
+        net.load_state_dict(chkpt["state_dict"])
+
+        if args.resume:
+            optimizer.load_state_dict(chkpt["optimizer"])
+            resume = chkpt["epoch"]
+
+    if model["opt"]["loss"] == "CrossEntropy":
+        criterion = CrossEntropyLoss2d(weight=weight).to(device)
+    elif model["opt"]["loss"] == "mIoU":
+        criterion = mIoULoss2d(weight=weight).to(device)
+    elif model["opt"]["loss"] == "Focal":
+        criterion = FocalLoss2d(weight=weight).to(device)
+    elif model["opt"]["loss"] == "Lovasz":
+        criterion = LovaszLoss2d().to(device)
+    else:
+        sys.exit("Error: Unknown [opt][loss] value !")
```

Comment:
Hm what's a good way to make this a bit more elegant? Can we e.g. put a `__name__` attribute in the loss classes or use the class name, and then lookup all available losses from a set of classes?

Reply:
It's quite easy indeed to retrieve all loss available, and pick up the good one if present.
But the point here is that there's no reason that all losses would share the same prototype (i.e lovasz one don't need weights as weighted crossentropy ones).

An idea to deal with that, while keeping the human readibility ?

Reply:
Should we create an interface to pass weights to all loss functions? And then write an adapter for lovasz which simply discards the weights and created a lovasz loss?

Reply:
In addition above we require the user to always calculate and pass in the dataset weights.

Should we still let the user do thiseven if the lovasz loss function doesn't need it?

Otherwise I think this pull request is ready to go - just this small nitpick here.

Reply:
Yes, good point !

Could been indeed improve to allow dataset weights optionnal if choosen loss don't need them...
72a13df commit handle this.

Reply:
As everything seems OK now, could you apply the PR ?
Thanks in advance :)